Release 2.1.0 (#2307)
nginxinc/kubernetes-ingress: docs/content/releases.md

toc: true
---

## NGINX Ingress Controller 2.1.0

05 Jan 2022

> 

Reviewer: We need to add an overview section.  and sorry, git is failing me this morning:

OVERVIEW

Support for NGINX App Protect Denial of Service protection with NGINX Ingress Controller.
More information about [NGINX App Protect DoS] https://www.nginx.com/products/nginx-app-protect/denial-of-service/

Full support for gRPC services using the NGINX Ingress Controller [VirtualServer and VirtualServerRoute] https://docs.nginx.com/nginx-ingress-controller/configuration/virtualserver-and-virtualserverroute-resources custom resource definitions.  This makes configuring and supporting gRPC services much easier, giving a simple YAML configuration and removing the need for snippets.
Resource definition examples for gRPC can be found [here] https://github.com/nginxinc/kubernetes-ingress/tree/release-2.1/examples/custom-resources/grpc-upstreams

Implementation of NGINX Plus persistent health checks in VirtualServer and VirtualServerRoute to further reduce interruptions to your service traffic as Ingress configuration changes happen in your dynamic Kubernetes environment.  NGINX Ingress Controller already supported mandatory health checks (not forwarding traffic until your upstreams were actually healthy).  Mandatory health checks can be marked as persistent, so that the previous state is remembered when reloading configuration.
See the settings [here] https://docs.nginx.com/nginx-ingress-controller/configuration/virtualserver-and-virtualserverroute-resources/#upstreamhealthcheck
More about the [NGINX Plus mandatory and persistent health check features] https://docs.nginx.com/nginx/admin-guide/load-balancer/http-health-check/#mandatory-health-checks

Reviewer 2: @brianehlert 

(1)
Would it make sense to add a reference to the doc for DOS?  It should be /nginx-ingress-controller/app-protect-dos/dos-protected/ (judging by this preview https://deploy-preview-2323--nginx-kubernetes-ingress.netlify.app/nginx-ingress-controller/app-protect-dos/dos-protected/ )

(2)
> NGINX Ingress Controller already supported mandatory health checks (not forwarding traffic until your upstreams were actually healthy).

both mandatory and persistent was introduced for VS. Mandatory was supported for health checks in Ingress resources (via an annotation), but that configuration is different from the configuration for health checks in VirtualServers.

I would also add a sentence from here (https://github.com/nginxinc/kubernetes-ingress/pull/2251) -- "When combined with the slow-start parameter, it gives a new server more time to connect to databases and “warm up” before being asked to handle their full share of traffic.", because mandatory is important for the slow start feature.

Reviewer: OVERVIEW

Support for NGINX App Protect Denial of Service protection with NGINX Ingress Controller.
More information about [NGINX App Protect DoS] https://www.nginx.com/products/nginx-app-protect/denial-of-service/
Examples for configuring NGINX App Protect DoS with NGINX Ingress Controller can be found [here] https://github.com/nginxinc/kubernetes-ingress/tree/release-2.1/examples/appprotect-dos

Full support for gRPC services using the NGINX Ingress Controller [VirtualServer and VirtualServerRoute] https://docs.nginx.com/nginx-ingress-controller/configuration/virtualserver-and-virtualserverroute-resources custom resource definitions.  This makes configuring and supporting gRPC services much easier, giving a simple YAML configuration and removing the need for snippets.
Resource definition examples for gRPC can be found [here] https://github.com/nginxinc/kubernetes-ingress/tree/release-2.1/examples/custom-resources/grpc-upstreams

Implementation of NGINX persistent health checks in VirtualServer and VirtualServerRoute to further reduce interruptions to your service traffic as configuration changes continuously happen in your dynamic Kubernetes environment(s).  NGINX Ingress Controller previously supported `mandatory` health checks (not forwarding traffic until your service pods were healthy) only with the Ingress resource. This has been extended to the Virtualserver and VirtualServerRoute custom resource definitions and augmented to include `persistent`.  
Mandatory health checks can be marked as persistent, so that the previous state is remembered when reloading configuration. When combined with the slow-start parameter, it gives a new service pod more time to connect to databases and “warm up” before being asked to handle their full share of traffic.
See the settings [here] https://docs.nginx.com/nginx-ingress-controller/configuration/virtualserver-and-virtualserverroute-resources/#upstreamhealthcheck
More about the [NGINX Plus mandatory and persistent health check features] https://docs.nginx.com/nginx/admin-guide/load-balancer/http-health-check/#mandatory-health-checks

Author: added the overview

FEATURES:
* [2251](https://github.com/nginxinc/kubernetes-ingress/pull/2251) Enable setting mandatory and persistent in upstream healthchecks in VS and VSR.
* [2241](https://github.com/nginxinc/kubernetes-ingress/pull/2241) Add support for NGINX App Protect DoS.
* [2200](https://github.com/nginxinc/kubernetes-ingress/pull/2200) Add Alpine image with OpenTracing.
* [2178](https://github.com/nginxinc/kubernetes-ingress/pull/2178) Support healthchecks in gRPC upstreams.

> * [2110](https://github.com/nginxinc/kubernetes-ingress/pull/2110) Support gRPC in the Upstreams of the virtual server resources.

Reviewer 3: Need to add a thanks here too to the original submitter.

Reviewer 3: ```suggestion
* [2110](https://github.com/nginxinc/kubernetes-ingress/pull/2110) Support gRPC in the Upstreams of the virtual server resources. Particular thanks to [Chiyu Zhong](https://github.com/CatTail) for all their work.
```

* [2100](https://github.com/nginxinc/kubernetes-ingress/pull/2100) Add support for initContainers. Thanks to [Gunnar Scherf](https://github.com/g10f).
* [1827](https://github.com/nginxinc/kubernetes-ingress/pull/1827) Add support for wildcard cert in VirtualServer resources. Thanks to [Simon Wachter](https://github.com/svvac).
* [2107](https://github.com/nginxinc/kubernetes-ingress/pull/2107) Add option to download the NGINX Ingress Controller binary. Introduced a new `TARGET` `download` in the `Makefile` which can be used when building the NGINX Ingress Controller Docker image. With this option the Ingress Controller binary will be downloaded instead of built from source.

IMPROVEMENTS:
* [2044](https://github.com/nginxinc/kubernetes-ingress/pull/2044) Upload NGINX Ingress Controller binaries to release.
* [2094](https://github.com/nginxinc/kubernetes-ingress/pull/2094) AP: update appolicies crd.
* [2216](https://github.com/nginxinc/kubernetes-ingress/pull/2216) Add grpc_status to the logs.
* [2237](https://github.com/nginxinc/kubernetes-ingress/pull/2237) Unbind app-protect from -preview-policies.

> * [2273](https://github.com/nginxinc/kubernetes-ingress/pull/2273) Make the resource comparison more informative in case of an error. [Andrey Karpov](https://github.com/ndk)

Reviewer 3: ```suggestion
* [2273](https://github.com/nginxinc/kubernetes-ingress/pull/2273) Make the resource comparison more informative in case of an error. Thanks to [Andrey Karpov](https://github.com/ndk)
```

toc: true
---

## NGINX Ingress Controller 2.1.0

06 Jan 2022

OVERVIEW:

> 

Reviewer 2: style: can we have a list here - this would give this section better structure? Like here https://docs.nginx.com/nginx-ingress-controller/releases/#nginx-ingress-controller-200

Author: @pleshakov pushed a commit to create a list, not sure I did it right 😅  feel free to fix it

Reviewer 2: @lucacome  looks allright! thanks!

Support for NGINX App Protect Denial of Service protection with NGINX Ingress Controller.
More information about [NGINX App Protect DoS](https://www.nginx.com/products/nginx-app-protect/denial-of-service/)
Examples for configuring NGINX App Protect DoS with NGINX Ingress Controller can be found [here](https://github.com/nginxinc/kubernetes-ingress/tree/release-2.1/examples/appprotect-dos)

Full support for gRPC services using the NGINX Ingress Controller [VirtualServer and VirtualServerRoute](https://docs.nginx.com/nginx-ingress-controller/configuration/virtualserver-and-virtualserverroute-resources) custom resource definitions.  This makes configuring and supporting gRPC services much easier, giving a simple YAML configuration and removing the need for snippets.

> Resource definition examples for gRPC can be found [here](https://github.com/nginxinc/kubernetes-ingress/tree/release-2.1/examples/custom-resources/grpc-upstreams)

Reviewer 3: This docs link should be updated to use the tag instead of the release branch.

> Implementation of NGINX persistent health checks in VirtualServer and VirtualServerRoute to further reduce interruptions to your service traffic as configuration changes continuously happen in your dynamic Kubernetes environment(s).  NGINX Ingress Controller previously supported `mandatory` health checks (not forwarding traffic until your service pods were healthy) only with the Ingress resource. This has been extended to the VirtualServer and VirtualServerRoute custom resource definitions and augmented to include `persistent`.

Reviewer 2: correction with some additional details
```suggestion
Implementation of NGINX mandatory and persistent health checks in VirtualServer and VirtualServerRoute to further reduce interruptions to your service traffic as configuration changes continuously happen in your dynamic Kubernetes environment(s). Health checks have been extended to include `mandatory` and `persistent` fields. Mandatory health checks ensures that a new upstream server starts receiving traffic only after the health check passes. Mandatory health checks can be marked as persistent, so that the previous state is remembered when the Ingress Controller reloads NGINX Plus configuration. When combined with the slow-start parameter, the mandatory health check give a new upstream server more time to connect to databases and “warm up” before being asked to handle their full share of traffic.
See the settings [here](https://docs.nginx.com/nginx-ingress-controller/configuration/virtualserver-and-virtualserverroute-resources/#upstreamhealthcheck)
More about the [NGINX Plus mandatory and persistent health check features](https://docs.nginx.com/nginx/admin-guide/load-balancer/http-health-check/#mandatory-health-checks)
```
